### PR TITLE
PP 193 expiring charges endpoint

### PR DESCRIPTION
**Pull request by rahul-vinayak**
- a new endpoint to expire the charges which are more than an hour old by default, allowed overriding them by env var
- service changes to have the logic of cancelling a charges before expiring charges when the status is AUTH_SUCCESS
- tests to test the service and the new end point.

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/model/ErrorResponse.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
+package uk.gov.pay.connector.model;
+
+import static uk.gov.pay.connector.model.ErrorType.*;
+
+public class ErrorResponse {
```
Comment (PauloPortugal) [inline src/main/java/uk/gov/pay/connector/model/ErrorResponse.java]: :+1:

File: src/main/java/uk/gov/pay/connector/model/ErrorType.java [hunk 2]
```diff
@@ -1,7 +1,8 @@
 package uk.gov.pay.connector.model;
 
-public enum GatewayErrorType {
+public enum ErrorType {
```
Comment (PauloPortugal) [inline src/main/java/uk/gov/pay/connector/model/ErrorType.java]: :+1:

File: src/main/java/uk/gov/pay/connector/model/GatewayError.java [hunk 3]
```diff
@@ -1,57 +0,0 @@
-package uk.gov.pay.connector.model;
-
-import static uk.gov.pay.connector.model.GatewayErrorType.*;
-
-public class GatewayError {
```
Comment (PauloPortugal) [inline src/main/java/uk/gov/pay/connector/model/GatewayError.java]: :+1:

File: src/main/java/uk/gov/pay/connector/service/CardService.java [hunk 4]
```diff
@@ -110,27 +114,35 @@ public CardService(GatewayAccountDao accountDao, ChargeDao chargeDao, PaymentPro
...
     }
 
-    public Either<GatewayError, GatewayResponse> doCancel(String chargeId, Long accountId) {
-        return chargeDao
-                .findByExternalIdAndGatewayAccount(chargeId, accountId)
-                .map(cancel())
+    public Either<ErrorResponse, GatewayResponse> doCancel(String chargeId, Long accountId) {
+        Optional<ChargeEntity> charge = chargeDao.findByExternalIdAndGatewayAccount(chargeId, accountId);
+
+        if(charge.isPresent() && hasStatus(charge.get(), EXPIRED)) {
```
Comment (PauloPortugal) [inline src/main/java/uk/gov/pay/connector/service/CardService.java]: :+1:

File: src/main/java/uk/gov/pay/connector/service/ChargeService.java [hunk 5]
```diff
@@ -63,6 +75,70 @@ public ChargeResponse create(Map<String, Object> chargeRequest, GatewayAccountEn
...
+        chargeEntities.stream().forEach(chargeEntity -> {
+            chargeEntity.setStatus(status);
+            chargeDao.mergeAndNotifyStatusHasChanged(chargeEntity);
+        });
+    }
+
+    public Map<String, Integer> expire(List<ChargeEntity> charges) {
+        Map<Boolean, List<ChargeEntity>> chargesToProcessExpiry = charges
+                .stream()
+                .collect(Collectors.partitioningBy(
```
Comment (PauloPortugal) [inline src/main/java/uk/gov/pay/connector/service/ChargeService.java]: Very clever guys! :)

**Comment by PauloPortugal** [comment]
@rahul-vinayak can you resolve the conflicts. Otherwise it's all good!

File: src/main/java/uk/gov/pay/connector/resources/ChargesResource.java [hunk 6]
```diff
@@ -62,6 +68,15 @@
     private ChargeDao chargeDao;
     private GatewayAccountDao gatewayAccountDao;
     private ChargeService chargeService;
+    private CardService cardService;
```
Comment (kakumara) [inline src/main/java/uk/gov/pay/connector/resources/ChargesResource.java]: is this `cardService` used anywhere in code?

File: src/main/java/uk/gov/pay/connector/resources/ChargesResource.java [hunk 7]
```diff
@@ -2,15 +2,22 @@
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
 import fj.F;
+import fj.data.Either;
```
Comment (kakumara) [inline src/main/java/uk/gov/pay/connector/resources/ChargesResource.java]: Quite a few unused imports

Comment (rahul-vinayak) [inline src/main/java/uk/gov/pay/connector/resources/ChargesResource.java]: sorted

File: src/main/java/uk/gov/pay/connector/resources/ChargesResource.java (hunk 6 above)
Comment (rahul-vinayak) [inline src/main/java/uk/gov/pay/connector/resources/ChargesResource.java]: removed, probably left while refactoring

File: src/main/java/uk/gov/pay/connector/resources/ChargesResource.java [hunk 8]
```diff
@@ -148,6 +163,26 @@ public Response createNewCharge(@PathParam("accountId") Long accountId, Map<Stri
                 .orElseGet(() -> notFoundResponse(logger, "Unknown gateway account: " + accountId));
     }
 
+    @POST
+    @Path(EXPIRE_CHARGES)
```
Comment (kakumara) [inline src/main/java/uk/gov/pay/connector/resources/ChargesResource.java]: [Suggestion] Wonder if its better to put this resource in its own resource and not mix up other payment flows... given its more of a backend job and there may be other jobs we may add on later on

Comment (rahul-vinayak) [inline src/main/java/uk/gov/pay/connector/resources/ChargesResource.java]: Ya we can definitely think of having a separate resource for this if we find the need of more such tasks, as of now we don't know if there would be, and this fits closest to the charge related resource. Would be 10 seconds to move it in future.

**Pull request merged**